src/daft-dsl/src/functions/mod.rs

```diff
@@ -118,6 +119,39 @@ pub fn function_display(f: &mut Formatter, func: &FunctionExpr, inputs: &[ExprRe
     Ok(())
 }
 
+pub fn function_display_without_formatter(
+    func: &FunctionExpr,
+    inputs: &[ExprRef],
+) -> std::result::Result<String, std::fmt::Error> {
+    let mut f = String::default();
```

Comment:
can we write it out directly to `std::fmt::Writer`?

Reply:
How so? I don't think `std::fmt::Writer` exists. Did you mean `std::fmt::Write`? That's a trait, however.

Reply:
Whoops I meant [Formatter](https://doc.rust-lang.org/beta/std/fmt/struct.Formatter.html)

I was wondering if instead of writing to a string (heap allocation) which then gets written to the Formatter, if we can write directly to the formatter

``
impl Display for X {
  fn fmt(&self, f: &mut Formatter) -> Result {}
}
```

Reply:
Hm. Not sure how to get access to the `Formatter` instance that the `fmt` function gets as its second argument. The docs say not to construct one yourself.

Reply:
Ya so it looks like we can't directly construct a `Formatter` instance, nor are we able to access the `Formatter` through the `#[display("{}", ...)]` macro. I was trying to do something like `#[display(f, "{}", ...)]`, but that wasn't working. Looks like `derive_more` doesn't want its users to access the underlying formatter directly.

Reply:
If we want access to the formatter, we'd need to revert back to the manual implementation of `Display` for `Expr`.